### PR TITLE
Speed up application startup by using `ld.so.cache`

### PR DESCRIPTION
**Pull request by winterqt**
###### Description of changes

This PR does two things:
1. Patch glibc's dynamic loader to load the ld.so cache from `$ORIGIN/../etc/ld.so.cache` instead of `/etc/ld.so.cache`, and to prefer the cache over `RPATH` entries
2. Introduce a new setup hook, `generate-ld-cache`, to facilitate the generation of these caches

This was originally implemented by Guix's Ludovic Courtès in August of 2021 (after Ricardo came up with the idea) -- see [their blog post](https://guix.gnu.org/en/blog/2021/taming-the-stat-storm-with-a-loader-cache/) and [their discussion](https://issues.guix.gnu.org/44899) on the change for more information + some benchmarks. The glibc patch is copied verbatim from the Guix codebase.

As noted in the linked Guix issue, there are a few things to consider here:
- The hard-coded `$ORIGIN/../etc/ld.so.cache` path means that it can only be used with first-level sub-directories like `bin` and `sbin`
- There is a possibility that users could be tricked into loading a malicious `ld.so.cache`
- If you read the discussion, you'll note that I omitted one of their points here: that transient dependencies aren't added to the cache. Our implementation actually *does include* transient dependencies, because we use `ldd` to get the required dependencies, as opposed to just using the data from the ELF.

When discussing this with pennae, they brought up the idea of embedding an absolute path to the cache in an ELF section, and adjusting the patch to do that. We may want to do this instead; I'd appreciate thoughts on the matter.

I'd like to thank the Guix folks mentioned above for originally coming up with and implementing this idea, as well as pennae for nerd sniping me into looking at perf stuff, and their thoughts on the matter.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by winterqt** [comment]
After looking around some more, I stumbled across @fzakaria'a [nix-harden-needed](https://github.com/fzakaria/nix-harden-needed), which may be a better choice over Guix's solution (per the README, they want to integrate it into Nixpkgs).

(Farid: hi, figured I'd loop you in here, considering you have interest in this problem :)

**Comment by fzakaria** [comment]
> After looking around some more, I stumbled across @fzakaria'a [nix-harden-needed](https://github.com/fzakaria/nix-harden-needed), which may be a better choice over Guix's solution (per the README, they want to integrate it into Nixpkgs).
> 
> (Farid: hi, figured I'd loop you in here, considering you have interest in this problem :)

Hi Winter!

Thank you for pinging me.
This is something I would love to solve at nixpkgs level -- I've just been intimated doing the solution although I've audited the code and have a good sense of what to do.

My hot-take:
1 - we can take https://github.com/fzakaria/shrinkwrap and rewrite it in C (it's using LIEF C-library) so the closure is very small and we can include it in the bootstrap binary.
     With shrinkwrap we can take a much simpler "patchelf" style approach and fixup the binaries.
2 - overtime do harden-needed approach to make it more "robust"
3 - I really want to investigate using $ORIGIN in the path so that the binaries are relocatable for any nix-store path.
(This is something Spack package manager accomplishes already)
(We actually need to upstream $ORIGIN support for PT_INTERP)

**Comment by fzakaria** [comment]
Here is a document of some other much more "visionary" ideas I have for Nix + NixOS.

[Breaking_Apart_the_Shared_Object_Monolith.pdf](https://github.com/NixOS/nixpkgs/files/10274134/Breaking_Apart_the_Shared_Object_Monolith.pdf)

**Comment by winterqt** [comment]
So I hope you don't mind me asking this here (if you'd rather discuss over e.g. IRC, Matrix, or email to avoid cluttering up this PR, let me know), but what does Shrinkwrap provide that this cache approach doesn't? They seemingly both eliminate the intensive RPATH traversal.

I know there's one advantage you list in the README:
> Secondly, shared dynamic objects may be found due to the fact that they are cached during the linking stage. Meaning, if another shared object requires the same dependency but failed to specify where to find it, it may still properly resolved if discovered earlier in the linking process. This is extremely error prone and changing any of the executable's dependencies can change the link order and potentially cause the binary to no longer work.

but I admit I'm confused as to what a practical example of this would look like/what the problem is in general 😅. Do you mind explaining?

**Comment by winterqt** [comment]
> Edit: This is a great PR btw -- very concise if it works. Only unfortunate thing is patching but maybe you can upstream something to make it configurable based on an ELF file section.

It does work; I've tested it with `hello`, and any Guix user has been using it since last year.

> It has the added benefit of supporting LD_LIBRARY_PATH which the other options don't (maybe though that's a feature?)

(Why doesn't Shrinkwrap support it, since it just lifts dependencies up to the parent ELF?)

**Comment by Mindavi** [comment]
Some questions:

- Does this work when crosa-compiling?
- Are there any implications for content-addressed derivations with this?

File: pkgs/stdenv/generic/default.nix [hunk 1]
```diff
@@ -71,7 +71,8 @@ let
       ../../build-support/setup-hooks/reproducible-builds.sh
       ../../build-support/setup-hooks/set-source-date-epoch-to-latest.sh
       ../../build-support/setup-hooks/strip.sh
-    ] ++ lib.optionals hasCC [ cc ];
+    ] ++ lib.optionals hostPlatform.isLinux [ ../../build-support/setup-hooks/generate-ld-cache.sh ]
```
Comment (raphaelr) [inline pkgs/stdenv/generic/default.nix:74]: hostPlatform.isLinux is no specific enough, for example pkgsMusl doesn't have a ldconfig. Maybe `hostPlatform.isLinux && hostPlatform.isGnu`?

```
generating LD cache for ELF executables in /nix/store/bgm2nivmwfa94pw6xlial071z9
0vv5pg-nano-7.1
found lib: /nix/store/62hdd7jfmj90wrwl6dp94hj6xzl9cgvq-ncurses-6.3-p20220507/lib
/libncursesw.so.6
found lib: /nix/store/mpgmlwjdd9g61w92q5d9g8y7r66fp8j6-musl-1.2.3/lib/ld-musl-x8
6_64.so.1
found lib dirs: /nix/store/62hdd7jfmj90wrwl6dp94hj6xzl9cgvq-ncurses-6.3-p2022050
7/lib /nix/store/mpgmlwjdd9g61w92q5d9g8y7r66fp8j6-musl-1.2.3/lib
/nix/store/6j095nxqj7qylwbbp4m00kb640gca06x-generate-ld-cache.sh: line 41: ldcon
fig: command not found
failed to generate LD cache
```
(nix-build -A pkgsMusl.nano)

**Comment by winterqt** [comment]
> buildEnv (and presumably symlinkJoin and friends) don't seem to like the approach :(

Adding `etc/ldconfig.so.cache` to this block will fix `buildEnv`:
https://github.com/NixOS/nixpkgs/blob/339ff0387aa51495586b69c9052e573465126bcb/pkgs/build-support/buildenv/builder.pl#L125-133

Not sure what to do about `symlinkJoin`, though; maybe it doesn't suffer from this issue (or at least won't fail to build)? Will test ASAP.

File: pkgs/build-support/setup-hooks/generate-ld-cache.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# shellcheck shell=bash
+
+fixupOutputHooks+=('if [ -z "${dontGenerateLDCache-}" ]; then generateLDCache "$prefix"; fi')
+
+generateLDCache() {
```
Comment (fzakaria) [inline pkgs/build-support/setup-hooks/generate-ld-cache.sh:5]: why not just use patchelf to list the dependencies rather than this bash ?
(does patchelf list the needed transitive? if not, you can write a tiny C binary using https://github.com/lief-project/LIEF)

Might be more maintainable.

File: pkgs/stdenv/generic/default.nix (hunk 1 above)
Comment (fzakaria) [reply pkgs/stdenv/generic/default.nix:74]: Good point this only works for Glibc

**Comment by tomberek** [comment]
There was a similar patch and discussion from https://github.com/NixOS/patchelf/pull/357#issuecomment-1030244165 that started to fold this into the patchelf phase. I'm a fan of the ld.so.cache approach due to the possibility of providing a different cache en-masse, or to mutable locations for testing/debugging!) (but the shrinkwrap also has merits.

Looks like the ldd searching for transitive deps is the main difference. I fear this makes the approach more fragile. What are the pros/cons here?

> When discussing this with pennae, they brought up the idea of embedding an absolute path to the cache in an ELF section, and adjusting the patch to do that. We may want to do this instead; I'd appreciate thoughts on the matter.

This would make it easier to then patch or redirect to a new place later. I can imagine a `patchelf --set-ld-so-cache /some-project/etc/ld.so.cache` or as a more focused alternative to lib.replaceDependency:

```
replaceLibDependency drv {
  drv = drv;
  from = openssl_old;
  to = openssl_new;
}
```

File: pkgs/development/libraries/glibc/dl-cache.patch [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
++ld_so_cache (void)
++{
++  static const char *loader_cache;
++
++  if (loader_cache == NULL)
++    {
++      static const char store[] = @STORE_DIRECTORY@;
++      const char *origin = _dl_get_origin ();
++
++      /* Check whether ORIGIN is something like "/gnu/store/…-foo/bin".  */
```
Comment (ncfavier) [inline pkgs/development/libraries/glibc/dl-cache.patch:29]:
```suggestion
+      /* Check whether ORIGIN is something like "/nix/store/…-foo/bin".  */
```

**Comment by nyabinary** [comment]
What the status on this currently?

**Comment by diniamo** [comment]
I don't understand, why aren't fzakaria's solutions plain better, compared to a cache?

**Comment by fzakaria** [comment]
Biggest wins seem to be if you are on NFS; I guess they are a minor (non-vocal) community :)

**Comment by LordGrimmauld** [comment]
Why? because of latency/bandwidth limit of network? In that case i'd expect some improvements for my external USB based setups too...

**Comment by fzakaria** [comment]
> Why? because of latency/bandwidth limit of network? In that case i'd expect some improvements for my external USB based setups too...

Yes latency; I wrote more in depth about it in https://arxiv.org/abs/2211.05118
tl;dr; you have n^2 stat calls to check if a library exists. Each stat call on NFS is a network call.
(Sometimes negative-cached which mitigates it but if not... 💥 )
